accepted/2025/nuget-mcp.md

```diff
@@ -0,0 +1,265 @@
+
+## Explanation
+
+### Improve MCP server authoring
+
+#### Today's experience
+
+To create a .NET MCP server today, the experience is somewhat self-guided and results in a NuGet package that looks no different from any other .NET CLI tool.
+
+An MCP server author can create a .NET console app, consume the [ModelContextProtocol NuGet package](https://www.nuget.org/packages/ModelContextProtocol), and implement tools that will be available to the LLM. The SDK helps with the stdio-based protocol that allows the end user's IDE to launch, discover, and invoke tools.
+
+The author will decide what CLI arguments or environment variables are needed to invoke the MCP server.
+
+When the MCP server is ready, the author will pack the [project as a tool with `<PackAsTool>`](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create) and publish the NuGet package to NuGet.org (or their private package feed), so that it can be used by other people.
+
+This approach works, but has a couple of problems:
+
+- The package has no indication that it is an MCP server. It just has the `DotnetTool` package type like any other .NET tool.
+  - NuGet.org and other package sources have no way of filtering packages to "MCP" or tailoring the browsing experience.
+- The package provides no guidance to the end user on how to start the MCP server.
+  - The end user doesn't have a good way of knowing the needed `command`, `args`, and `env` settings they should use.
+- The package author needs to set up their project just right, with a console app, .NET tool, and MCP dependency.
+
+#### New package type
+
+To address the filtering and discoverability problem, we will introduce an additional NuGet package type of `McpServer`.
+
+Currently, .NET tools get the `DotnetTool` package type, defined in the resulting package manifest (.nuspec file) generated by `dotnet pack`. [As of .NET 10](https://github.com/dotnet/sdk/pull/48039), it is possible to add additional package types to .NET tool packages using the `<PackageType>McpServer</PackageType>` MSBuild property in the tool .csproj.
+
+The resulting MCP server .nupkg will have two package types: `DotnetTool` and `McpServer`. For more information on package types, [see NuGet documentation](https://learn.microsoft.com/en-us/nuget/reference/nuspec#packagetypes).
+
+#### Startup instructions
+
+We have two options to encode the startup instructions into the package:
+
+1. Instruct the package author to include the desired consumer MCP JSON in the readme, and allow NuGet.org to scrape the JSON from the README markdown (a code block matching a certain pattern).
```

Comment:
Why would nuget.org need to scrape the JSON? It already presents the README, which the user sees and can copy from and paste into their mcp.json file.

Reply:
This would make the command palette on nuget.org MCP aware so it's clearer how to "install" the thing. The command palette for .NET tools would just be to install it, not run it (let alone run it with needed args/env).

Reply:
I imagine we won't make the consumption json mandatory for the publishers, so we won't be able to show it for all MCP packages. In cases where it's missing, is there a default we can show? If so, what would it look like?

Reply:
I think a default would cause more confusion than do good if it wasn't guaranteed to work. To guarantee it would work, you'd need to know the basic command line to use as well as any switches and env vars that are required (which will be common for MCP servers that access network resources).

Reply:
> In cases where it's missing, is there a default we can show?

I think we would show a user message not a JSON configuration or something meant to be machine readable. Something like "This NuGet package contains an MCP server. Check the package README and project documentation for details on how to launch the MCP server." 

The command palette tabs for global and local tools would still be there.

We could show a warning to the author (not the end user) indicating that the `server.json` is missing, much like we do when a package is missing a README.md.